### ISSUE TITLE
detox: release 0.11

### ISSUE THREAD

**Issue by ebertti** [tox-dev/tox#500]
Just a question?

When you gonna release version 0.11 with __main__.py on it?

**Comment by obestwalter** [comment]
No idea. Is anything stopping us to release this?

**Comment by hpk42** [comment]
I've uploaded and tested detox-0.11 at https://devpi.net/hpk/dev 
but couldn't run any windows tests yet ... can someone?

holger

**Comment by The-Compiler** [comment]
I just did a run on Windows, and ran into [some problems](https://devpi.net/hpk/dev/detox/0.11/+toxresults/detox-0.11.tar.gz/detox-0.11.tar.gz.toxresult2):

- `multiprocessing.cpu_count()` fails with `NotImplementedError` as it can't get the CPU count (added in https://github.com/tox-dev/detox/commit/c368cef5f3a1cf3af7da8d258068627c3a00ae00, cc @Pawamoy)
- Some timeouts in `self.greenlet.switch()`, but I have no idea what I'm looking at there

**Comment by hpk42** [comment]
On Mon, Apr 03, 2017 at 02:00 -0700, Florian Bruhin wrote:
> I just did a run on Windows, and ran into [some problems](https://devpi.net/hpk/dev/detox/0.11/+toxresults/detox-0.11.tar.gz/detox-0.11.tar.gz.toxresult2):
> 
> - `multiprocessing.cpu_count()` fails with `NotImplementedError` as it can't get the CPU count (added in https://github.com/tox-dev/detox/commit/c368cef5f3a1cf3af7da8d258068627c3a00ae00, cc @Pawamoy)

I just worked around it by falling back to "2" as default for the "-n" number of processes
parameter ...

> - Some timeouts in `self.greenlet.switch()`, but I have no idea what I'm looking at there

me neither -- i guess it's related to py36/eventlet compatibility and i am inclined
to ignore it ...

**Comment by Pawamoy** [comment]
- `multiprocessing.cpu_count()` fails with `NotImplementedError`: as @hpk42 said, we can fix a value, such as `2`, or use Windows compliant `psutil.cpu_count()` but I'm not sure introducing a new dependency is a good idea.

- Some timeouts in `self.greenlet.switch()`: if it comes from the functional tests (testing that a run with n processes is shorter than with n-1 processes), maybe we can remove them because these tests rely too much on the machine they are executed: spawning 2 processes in parallel on one CPU is not obviously faster than 2 processes sequentially. Spawning too much parallel processes can even be slower (am I right?).

For py36/eventlet compatibility, maybe see @hpk42 's [comment here](https://github.com/tox-dev/tox/issues/439#issuecomment-273167874) about merging detox into tox: it could be rewritten using plain threads to get rid of eventlet, greenlet and enum-compat dependencies, but it's not in the scope of this issue.

**Comment by hpk42** [comment]
ok, i released detox-0.11 despite the eventlet/windows failure -- let's see if it causes any real issues for people.

@Pawamoy i think it'd be good to try move detox to use threads, release that and if it works see to integrate it into tox proper.  Happy if you or someone else tried that, it shouldn't be hard (famous last ...).

**Comment by Pawamoy** [comment]
@The-Compiler Just had access to a Windows 7 machine, following is the output of the tests.

First, with only Python2.7 installed: failure, because env names "py1" and "py2" confuse tox, see `Captured stdout call` section.

```
~\Documents\GitHub\detox [master =]> tox
GLOB sdist-make: C:\Users\Pawamoy\Documents\GitHub\detox\setup.py
py27 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py27
py27 installdeps: pytest, tox
py27 inst: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\dist\detox-0.11.zip
py27 installed: appdirs==1.4.3,colorama==0.3.7,detox==0.11,enum-compat==0.0.2,enum34==1.1.6,eventlet==0.21.0,greenlet==0.4.12,packaging==16.8,pluggy==0.4.0,py==
1.4.33,pyparsing==2.2.0,pytest==3.0.7,six==1.10.0,tox==2.7.0,virtualenv==15.1.0
py27 runtests: PYTHONHASHSEED='88'
py27 runtests: commands[0] | py.test --junitxml=C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py27\log/junit-py27.xml
============================= test session starts =============================
platform win32 -- Python 2.7.13, pytest-3.0.7, py-1.4.33, pluggy-0.4.0
rootdir: C:\Users\Pawamoy\Documents\GitHub\detox, inifile:
collected 12 items

tests\test_detox.py ..........F
tests\test_main.py .

 generated xml file: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py27\log\junit-py27.xml
================================== FAILURES ===================================
______________________ TestProcLimitOption.test_runtests ______________________

self = <test_detox.TestProcLimitOption instance at 0x00000000039964C8>
cmd = <conftest.Cmd instance at 0x0000000003996888>

    @pytest.mark.timeout(60)
    def test_runtests(self, cmd):
        now1 = datetime.now()
        cmd.rundetox("-n", "1", "-epy1,py2")
        then1 = datetime.now()
        delta1 = then1 - now1
        assert delta1 >= timedelta(seconds=2)

        now2 = datetime.now()
        cmd.rundetox("--num", "2", "-epy1,py2")
        then2 = datetime.now()
        delta2 = then2 - now2
        assert delta2 >= timedelta(seconds=1)

>       assert delta1 >= delta2, 'pool size=2 took much time than pool size=1'
E       AssertionError: pool size=2 took much time than pool size=1
E       assert datetime.timedelta(0, 12, 698000) >= datetime.timedelta(0, 12, 917000)

C:\Users\Pawamoy\Documents\GitHub\detox\tests\test_detox.py:134: AssertionError
---------------------------- Captured stdout setup ----------------------------
example3 created at c:\users\pawamoy\appdata\local\temp\pytest-0\test_runtests1\example3
---------------------------- Captured stdout call -----------------------------
running ['C:\\Users\\Pawamoy\\Documents\\GitHub\\detox\\.tox\\py27\\Scripts\\detox.EXE', '-n', '1', '-epy1,py2'] curdir= c:\users\pawamoy\appdata\local\temp\pytest-0\test_runtests1\example3
py1 create: c:\users\pawamoy\appdata\local\temp\pytest-0\test_runtests1\example3\.tox\py1
py1 runtests: PYTHONHASHSEED='54'
py1 runtests: commands[0] | python -c import time; time.sleep(1)
py2 create: c:\users\pawamoy\appdata\local\temp\pytest-0\test_runtests1\example3\.tox\py2
ERROR: InterpreterNotFound: python2
___________________________________ summary ___________________________________
  py1: commands succeeded
ERROR:   py2: InterpreterNotFound: python2
running ['C:\\Users\\Pawamoy\\Documents\\GitHub\\detox\\.tox\\py27\\Scripts\\detox.EXE', '--num', '2', '-epy1,py2'] curdir= c:\users\pawamoy\appdata\local\temp\pytest-0\test_runtests1\example3
py1 create: c:\users\pawamoy\appdata\local\temp\pytest-0\test_runtests1\example3\.tox\py1
py2 create: c:\users\pawamoy\appdata\local\temp\pytest-0\test_runtests1\example3\.tox\py2
ERROR: InterpreterNotFound: python2
py1 runtests: PYTHONHASHSEED='109'
py1 runtests: commands[0] | python -c import time; time.sleep(1)
___________________________________ summary ___________________________________
  py1: commands succeeded
ERROR:   py2: InterpreterNotFound: python2

=========== 1 failed, 11 passed, 5 pytest-warnings in 76.88 seconds ===========
ERROR: InvocationError: 'C:\\Users\\Pawamoy\\Documents\\GitHub\\detox\\.tox\\py27\\Scripts\\py.test.EXE --junitxml=C:\\Users\\Pawamoy\\Documents\\GitHub\\detox\\.tox\\py27\\log/junit-py27.xml'
py34 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py34
ERROR: InterpreterNotFound: python3.4
py35 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py35
ERROR: InterpreterNotFound: python3.5
py36 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py36
ERROR: InterpreterNotFound: python3.6
__________________________________ summary ___________________________________
ERROR:   py27: commands failed
ERROR:   py34: InterpreterNotFound: python3.4
ERROR:   py35: InterpreterNotFound: python3.5
ERROR:   py36: InterpreterNotFound: python3.6
```

To fix this, I renamed "py1" and "py2" in "env1" and "env2" in `conftest.py`, example 3.

```
~\Documents\GitHub\detox [master = +0 ~2 -0 !]> tox
GLOB sdist-make: C:\Users\Pawamoy\Documents\GitHub\detox\setup.py
py27 inst-nodeps: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\dist\detox-0.11.zip
py27 installed: appdirs==1.4.3,colorama==0.3.7,detox==0.11,enum-compat==0.0.2,enum34==1.1.6,eventlet==0.21.0,greenlet==0.4.12,packaging==16.8,pluggy==0.4.0,py==1.4.33,pyparsing==2.2.0,pytest==3.0.7,six==1.10.0,tox==2.7.0,virtualenv==15.1.0
py27 runtests: PYTHONHASHSEED='909'
py27 runtests: commands[0] | py.test --junitxml=C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py27\log/junit-py27.xml
============================= test session starts =============================
platform win32 -- Python 2.7.13, pytest-3.0.7, py-1.4.33, pluggy-0.4.0
rootdir: C:\Users\Pawamoy\Documents\GitHub\detox, inifile:
collected 12 items

tests\test_detox.py ...........
tests\test_main.py .

 generated xml file: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py27\log\junit-py27.xml

================ 12 passed, 5 pytest-warnings in 89.48 seconds ================
py34 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py34
ERROR: InterpreterNotFound: python3.4
py35 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py35
ERROR: InterpreterNotFound: python3.5
py36 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py36
ERROR: InterpreterNotFound: python3.6
__________________________________ summary ___________________________________
  py27: commands succeeded
ERROR:   py34: InterpreterNotFound: python3.4
ERROR:   py35: InterpreterNotFound: python3.5
ERROR:   py36: InterpreterNotFound: python3.6
```

Then, after installing Python3.5 and Python3.6 (no executable installer for Python3.4/Windows):

```
~\Documents\GitHub\detox [master = +0 ~2 -0 !]> detox
GLOB sdist-make: C:\Users\Pawamoy\Documents\GitHub\detox\setup.py
py34 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py34
ERROR: InterpreterNotFound: python3.4
py35 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py35
py36 create: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py36
py27 inst-nodeps: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\dist\detox-0.11.zip
py27 runtests: PYTHONHASHSEED='438'
py27 runtests: commands[0] | py.test --junitxml=C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py27\log/junit-py27.xml
py35 installdeps: pytest, tox
py36 installdeps: pytest, tox
py36 inst: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\dist\detox-0.11.zip
py35 inst: C:\Users\Pawamoy\Documents\GitHub\detox\.tox\dist\detox-0.11.zip
py35 runtests: PYTHONHASHSEED='438'
py35 runtests: commands[0] | py.test --junitxml=C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py35\log/junit-py35.xml
py36 runtests: PYTHONHASHSEED='438'
py36 runtests: commands[0] | py.test --junitxml=C:\Users\Pawamoy\Documents\GitHub\detox\.tox\py36\log/junit-py36.xml
__________________________________ summary ___________________________________
  py27: commands succeeded
ERROR:   py34: InterpreterNotFound: python3.4
  py35: commands succeeded
  py36: commands succeeded
```

As you can see, no `NotImplementedError` was raised, although I think `multiprocessing.cpu_count()` detected only 3 CPUs instead of 4. 

@hpk42 If nobody assigns to it, I will give a try to using plain threads for detox (when I find some time!!) :ghost:

**Comment by obestwalter** [comment]
Pretty sure this failure is due to Windows needing NUMBER_OF_PROCESSORS to be passed through - this was just fixed in tox and will be part of 2.8: https://github.com/tox-dev/tox/issues/517
